.emacs.d/config.org

```diff
@@ -55,62 +55,392 @@
+
+  #+BEGIN_SRC emacs-lisp
+  (use-package prettier-js
+    :hook ((js-mode typescript-mode rjsx-mode) . prettier-js-mode))
+  #+END_SRC
+
+  [[https://github.com/ananthakumaran/tide][TypeScript Interactive Development Environment for Emacs]]
+
+  #+BEGIN_SRC emacs-lisp
+  ;; TODO clean this up?
+  (defun setup-tide-mode ()
+    "Setup tide for javascript."
+    (interactive)
+    (tide-setup)
+    (flycheck-mode +1)
+    (setq flycheck-check-syntax-automatically '(save mode-enabled))
+    (eldoc-mode +1)
+    (tide-hl-identifier-mode +1)
+    (company-mode +1))
+
+  (use-package tide
+    :after
+    (typescript-mode company flycheck)
+    :hook
+    ((typescript-mode . tide-setup)
+     (typescript-mode . tide-hl-identifier-mode)
+     (before-save . tide-format-before-save))
+    :config
+    (flycheck-add-next-checker 'typescript-tide 'javascript-eslint)
+    (flycheck-add-next-checker 'tsx-tide 'javascript-eslint))
+  #+END_SRC
+
+  Modern React, note it’s generally a bad idea to bind this to ts or tsx since TypeScript uses =:= for types, which rjsx doesn’t seem to interpret well (it would be very nice if this was fixed!):
+
+  #+BEGIN_SRC emacs-lisp
+  (use-package rjsx-mode
+    :mode
+    (("\\.js\\'"   . rjsx-mode)
+     ("\\.jsx\\'"  . rjsx-mode)
+     ("\\.json\\'" . javascript-mode))
+    :magic ("/\\*\\* @jsx React\\.DOM \\*/" "^import React")
+    :init
+    (setq-default rjsx-basic-offset 2)
+    (setq-default rjsx-global-externs '("module" "require" "assert" "setTimeout" "clearTimeout" "setInterval" "clearInterval" "location" "__dirname" "console" "JSON")))
+
+  (use-package react-snippets
+    :after yasnippet)
+  #+END_SRC
+
+  For vue specifics as well:
+
+  #+BEGIN_SRC emacs-lisp
+  (use-package vue-html-mode)
+
+  (use-package vue-mode
+    :mode
+    (("\\.vue\\'"  . vue-mode)))
+  #+END_SRC
+
+** Java
+
+   Java is a built-in mode, but its LSP and DAP interfaces are included separately from the main =lsp-mode= and =dap-mode=.
+
+  #+BEGIN_SRC emacs-lisp
+  (use-package lsp-java
+    :hook
+    ((java-mode . lsp)
+     (java-mode . (lambda () (require 'dap-java)))))
+  #+END_SRC
+
+** Python
+
+   Unfortunately the Python package situation is a little confusing. There are a few older separate packages that provide a Python major mode, but we prefer the built-in one and avoid downloading a third-party one by not ensuring it. This allows us to have =use-package= configure Python mode without downloading anything.
+
+   #+BEGIN_SRC emacs-lisp
+   (use-package python
+     :ensure nil
+     :custom
+     (python-indent-guess-indent-offset-verbose nil)
+     (python-fill-docstring-style 'pep-257-nn))
+   #+END_SRC
+
+   To use a language server with Python, there are two options:
+   - =pyls= from [[https://github.com/palantir/python-language-server][Palantir]]: =pip install python-language-server=, and =lsp-mode= will automatically pick it up
```

Comment:
If we remove the hooks from the lsp definition above, we will need `  :hook (python-mode . lsp)`

Reply:
Done in d3c55ee70b3d72754b39312ba18ec5d1f09cb798